x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeParseProcessor.java

```diff
@@ -6,86 +6,110 @@
 
-    public static final String NAME = "dtparse";
+    public enum Parser {
+        DATE_TIME((timestampStr, pattern) -> DateTimeFormatter.ofPattern(pattern, Locale.ROOT)
+                .parseBest(timestampStr, ZonedDateTime::from, LocalDateTime::from)),
+        TIME((timestampStr, pattern) -> DateTimeFormatter.ofPattern(pattern, Locale.ROOT)
+                .parseBest(timestampStr, OffsetTime::from, LocalTime::from));
 
-    public DateTimeParseProcessor(Processor source1, Processor source2) {
-        super(source1, source2, null);
-    }
+        private final BiFunction<String, String, TemporalAccessor> parser;
 
-    public DateTimeParseProcessor(StreamInput in) throws IOException {
-        super(in);
-    }
-
-    /**
-     * Used in Painless scripting
-     */
-    public static Object process(Object timestampStr, Object pattern) {
-        if (timestampStr == null || pattern == null) {
-            return null;
-        }
-        if (timestampStr instanceof String == false) {
-            throw new SqlIllegalArgumentException("A string is required; received [{}]", timestampStr);
-        }
-        if (pattern instanceof String == false) {
-            throw new SqlIllegalArgumentException("A string is required; received [{}]", pattern);
+        Parser(BiFunction<String, String, TemporalAccessor> parser) {
+            this.parser = parser;
         }
 
-        if (((String) timestampStr).isEmpty() || ((String) pattern).isEmpty()) {
-            return null;
-        }
+        public Object parse(Object timestamp, Object pattern) {
+            if (timestamp == null || pattern == null) {
+                return null;
+            }
+            if (timestamp instanceof String == false) {
+                throw new SqlIllegalArgumentException("A string is required; received [{}]", timestamp);
+            }
+            if (pattern instanceof String == false) {
+                throw new SqlIllegalArgumentException("A string is required; received [{}]", pattern);
+            }
 
-        try {
-            TemporalAccessor ta = DateTimeFormatter.ofPattern((String) pattern, Locale.ROOT)
-                .parseBest((String) timestampStr, ZonedDateTime::from, LocalDateTime::from);
-            if (ta instanceof LocalDateTime) {
-                return ZonedDateTime.ofInstant((LocalDateTime) ta, ZoneOffset.UTC, UTC);
-            } else {
-                return ta;
+            if (((String) timestamp).isEmpty() || ((String) pattern).isEmpty()) {
+                return null;
             }
-        } catch (IllegalArgumentException | DateTimeException e) {
-            String msg = e.getMessage();
-            if (msg.contains("Unable to convert parsed text using any of the specified queries")) {
-                msg = "Unable to convert parsed text into [datetime]";
+            try {
+                TemporalAccessor ta = parser.apply((String) timestamp, (String) pattern);
```

Comment:
You should distinguish based on the enum value, this way you can have a specialized message for each and not a generic one that refers to both datetime and time.

Reply:
See my code changes, is it ok?